docs: add first draft of Sub-Optimal Provider APIs for Wallets CPS (#3)
will-break-it/wallet-architecture: docs/CPS/CPS-XXXX/README.md

## Problem
> [!NOTE]
>
> Fundamental wallet functionality is defined in this document as being able to construct and submit any transaction supported by the latest Cardano protocols, requiring:
> - **UTxO set** owned by addresses with credentials controlled by the wallet, that are either derived within the 1852 purposed, [BIP32-ed25519](https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf) accounts or the product of hashing a script.

> > - **Active stake pools**: to verify a nominated Stake Pools ID is valid and active at the time of delegation.

Reviewer: I'm not sure this is required in base API :thinking: Without metadata/metrics, wallet can't really help the user to choose a stake pool. Yes this could validate it's 'active', but users would have to use an external source anyway. 
```suggestion
```

Author: The idea is to verify a delegation is legitimate, rather than helping the user choose the pool. Now I think about it, we would also need to include DReps if this is the thinking.

Reviewer: I understand, but I don't think it's something that wallets definitely have to do. If node allows delegating to an unexisting/inactive stake pool (I believe it does but not 100% sure), then some wallets might want to allow that as well. A likely use case is for registering a stake pool and delegating to it before it becomes active.

Author: Ah, good point. I will remove the "active" aspect of it, leaving the responsibility to just a registered entity

Author: https://github.com/will-break-it/wallet-architecture/pull/3/commits/8a66ba20bef38aab4583e64af1c82da76b7b63ba

Created: 2024-07-09
License: CC-BY-4.0
---

## Abstract
The event-based nature of a Blockchain is well-suited for wallets to derive their internal state from genesis, however, most data services only expose a request/response interface and/or take on the responsibility of decoding the native binary data into JSON, enforcing a strong client/server contract with the current protocol transaction format. Wallets must use long polling to detect changes, on an interval that balances latency with network and service demand, then make numerous calls against various endpoints that have the potential to change on each protocol upgrade. Furthermore, the lack of an API standard inhibits development and adoption of a standalone desktop service for wallets to offer a full node experience for their users, which is best achieved via a universal API running either on a remote server, in a multi-tenant mode, or locally running in single-tenant mode.

## Problem
> [!NOTE]
>
> Fundamental wallet functionality is defined in this document as being able to construct and submit any transaction supported by the latest Cardano protocols, requiring:

> > - UTxO set owned by addresses derived within the 1852 purposed, [BIP32-ed25519](https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf) accounts or the product of hashing a script.

Reviewer: Would be great to clarify/rephrase this a little bit
```suggestion
> - UTxO set owned by addresses with credentials controlled by the wallet, that are either
- derived within the 1852 purposed, [BIP32-ed25519](https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf) accounts
- the product of hashing a script.
```

Author: https://github.com/will-break-it/wallet-architecture/pull/3/commits/4599ee91608add1bb16e08f92b069283e0f1ec08

> - Active stake pools
> - Reward account balances
> - Certificates submitted by the wallet
> - Current Cardano network protocol parameters

> > - Era summaries, for time conversions

Reviewer: Not strictly required for fundamental functionality; If we add this, then might as well add entire tx history instead of just certificates submitted by the wallet
```suggestion
```

Author: https://github.com/will-break-it/wallet-architecture/pull/3/commits/3f074c21d7f7748213637511c668634badb36562

> > - Current chain tip

Reviewer: Would be great to elaborate on use cases of each type of data
```suggestion
> - Current chain tip: for setting validity interval and/or creating time-locked scripts
```

Author: https://github.com/will-break-it/wallet-architecture/pull/3/commits/253da2d1a10bf119c2484246bced8c71243988f5

### Impedance Mismatch

> - The current APIs available to light wallets tend to be part of a general purpose solution serving many application types, built on top of [Cardano DB Sync] with a request/response interaction style. To achieve a reactive UX and stay in sync with the network state, long polling is required to reconstruct the original event-based nature of the system.

Reviewer: Not sure I understand this correctly, please elaborate :thinking: I believe many wallets are not re-constructing, but fetching current state that was re-constructed in the backend
```suggestion
- The current APIs available to light wallets tend to be part of a general purpose solution serving many application types, built on top of [Cardano DB Sync] with a request/response interaction style. To achieve a reactive UX and stay in sync with the network state, long polling is required to fetch the current state of the system and check whether there are any updates.
```

Author: https://github.com/will-break-it/wallet-architecture/pull/3/commits/d188be9a67146d3724a43dca709040aa376ef27b

### Fragile API 
- The API contract between a wallet and most data providers is unnecessarily complex, increasing the maintenance burden and transmission footprint. 

> [!WARNING]
> 
> This fragility in the API creates friction around each protocol upgrade, with wallets needing to synchronise development effort, which often goes through multiple interations over the course of months. Client libraries are readily available to deserialize CBOR

### Lacking a Standard
- This complicates or inhibits products offering a full node experience, with a free and open source service, available for users to install then choose their wallet of choice.

## Use cases

### Lace Extension

#### What is it doing?
- **Long-polling the tip for changes**. An HTTP GET request is made to a webserver every 5 seconds, checking if there's a new tip in the [Cardano DB Sync] database. If so, requests are then made to fetch transactions that are associated with the provided list of addresses.
- Transaction data in responses are **transmitted as JSON**, via a **custom schema**

> - Data is sourced from the combination of [Cardano DB Sync] and custom projections into the PostgreSQL DB. 

Reviewer: ```suggestion
- Data is sourced from the combination of [Cardano DB Sync], Ogmios LSQs and custom projections into the PostgreSQL DB. 
```

Author: https://github.com/will-break-it/wallet-architecture/pull/3/commits/c02e163b7bec75b4f3ecf7c40c78ef8f747b823f

### Impedance Mismatch
- The current APIs available to light wallets tend to be part of a general purpose solution serving many application types, built on top of [Cardano DB Sync] with a request/response interaction style. To achieve a reactive UX and stay in sync with the network state, long polling is required to detect progression of the network state and then check if there are any updates relevant to the wallet. i.e. It takes more work than necessary to maintain the event-based nature of the system.

> [!NOTE]
> 
> - Wallets require a local node to create a trustless connection with the network for fundamental behaviour, however, the current node technology renders it impractical for most users.
> - While current technology is a limiting aspect today, we can use this as inspiration and guidance to determine a solution that satisfies some of the desirable properties.

### Fragile API 
- The API contract between a wallet and most data providers is unnecessarily complex, increasing the maintenance burden and transmission footprint.
- Wallets can easily deserialize CBOR, using readily available client libraries.

> - Using Mithril to validate data requires an agreed standard 

Reviewer: Is the idea here that we need CBOR in order to compute the hash to validate transaction body?

Author: Yes

## Goals
- Define a standard API that is optimal for fundamental wallet operation. Consider extensibility for optional enhancement.
- Deliver data as a filtered subset of the network state, using native encoding formats, to simplify server implementations and push processing to the edge of the network.
- Foster long term stability of the API when crossing major Cardano protocol upgrade boundaries.
- Facilitate event-driven wallet behavior, minimising the need for long-polling, and to increase the sophistication of infrastructure management and user-base insights.
- Support script-based wallets
- Minimise the network demand by allowing the server to control the throttling strategy, rather than a pull-based approach like we see in the local chain-sync protocol. 
- API can be applied in a local, single-tenant environment, and a remote multi-tenant deployment, to facilitate a universal approach to wallet supporting an optional full local node for power users. 
- Remove the need for wallets to perform address discovery

> - Make it possible for a user to quickly derive new accounts and their addresses without re-projecting from genesis. 

Reviewer: How can we achieve this on `account` level?

Author: Convention of discovering addresses on index 0 + 1, then expand to 2 as we find on 1, so on and so forth. Having said this I can see we would then need to make it possible for the API to accept either an extended pub key or account pub key (for more privacy) :thinking:

Reviewer: Subscribing with account pub key is a reasonable option, but it can only be used to derive payment/staking keys for that account - wouldn't be able to discover other accounts. Also for script wallets, we'd still need the credentials (or script hashes).

I believe account discovery also wouldn't work if we wanted to be subscribing with root pub key because of the 'hardening'. That is why we needed to add password prompt for enabling accounts when we implemented multi-account.

Author: Ah good point. I might just drop this out of the document, as again, it's more of a concern relating to a single-tenant  implementation.

Author: https://github.com/will-break-it/wallet-architecture/pull/3/commits/da24badbf4b5c0abf5f75f2323842829afef4fdd

## Open Questions
- What identifiers will wallets provide for the server to filter on?
- What is the most efficient solution for indexing data based on account identifiers?
- Will it support multiple accounts?
- How will chain rollbacks be handled?
- What design considerations are made to establish a stable API over network protocol upgrades?
- Will the API standard support extensions?
- Should verification of stake and voting rights delegation be considered part of fundamental wallet behaviour or an extension?
- Can Mithril be used to verify state snapshots for provided account credentials, to facilitate a fast catchup feature? If not, what is a fallback approach for resource and/or network-constrained applications?
- How will the API behaviour differ for single-tenant use cases?
- Are there any natural authentication strategies?
- Should the Block be transmitted to wallet, similar to the Ouroboros Chain-Sync protocol, or just the relevant transactions?   
- Will this API be applicable for other Blockchains?

> - What is the source of rewards data, given there is no node API currently?

Reviewer: Is this really a problem for base API? There is a LSQ to get reward account summaries.

Detailed reward data is required for
- displaying wallet's reward history
- stake pool metrics

Which are nice to have but not essential

Author: I'll reword this. My intention was to setup a question we have the answer to already

Author: Just removed it instead https://github.com/will-break-it/wallet-architecture/pull/3/commits/27802ddb1577fd1fc8657421ec6fb7ed01d1c058

Author: Just removed it instead https://github.com/will-break-it/wallet-architecture/commit/7b7ca0ce64ace467ca8e09a0b7d722c9f91e371d

> - What is the most efficient strategy to prepare the database for single-tenant use-cases?

Reviewer: This seems like a client-side concern, while the rest of the CPS is about a backend API

Author: The database is a server concern, but I agree this is more of an implementation concern. Will remove

Author: https://github.com/will-break-it/wallet-architecture/pull/3/commits/27802ddb1577fd1fc8657421ec6fb7ed01d1c058

Author: https://github.com/will-break-it/wallet-architecture/pull/3/commits/7b7ca0ce64ace467ca8e09a0b7d722c9f91e371d

Author: https://github.com/will-break-it/wallet-architecture/commit/27802ddb1577fd1fc8657421ec6fb7ed01d1c058

### Inefficient Address Discovery
- Scanning a [BIP-44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#account-discovery) keychain for used addresses has the potential to be quite an intensive operation and is non-trivial to implement in the wallet.
- Most APIs support query by address only, missing the simplified alternative approach of using payment and stake key credentials. Funds intentionally or inadvertently sent to an address that would otherwise not be discovered under BIP-44 discovery standards would be found with this approach.
- Funds under the control of unknown [Franken Address](https://www.essentialcardano.io/glossary/franken-address) go undetected, despite the wallet having part ownership.

## Use cases

### Lace Extension

#### What is it doing?
- Transaction data in responses are **transmitted as JSON**, via a **custom schema**
- Data is sourced from the combination of [Cardano DB Sync], Ouroboros Local State Queries, and custom Ouroboros Chain-Sync projections into the PostgreSQL DB.

> - **Polling the tip for changes**. An HTTP GET request is made to a webserver every 5 seconds, checking if there's a new tip in the [Cardano DB Sync] database. If so, requests are then made to fetch transactions that are associated with the provided list of addresses after a certain point.

Reviewer 2: I remember discussing why we must fetch the tip every 5 seconds, something related to NFTs? I suggest we note it in this document since it is the most frequently made call.

Author: The polling rate was selected to just optimise the UX, rather than a specific use case, since we have sending statuses to update etc. We could back this off to 10 or more seconds without too much impact

#### Why?
- The philosophy in Lace is to place the responsibility of deriving wallet state from the historical transactions, similar to projecting state via the chain-sync protocol.
- Blockfrost was used during the early development of the Lace frontend, however the design of the API to require many small requests conflicts with the rollback handling strategy of Lace, and therefore was not a good fit for fetching transactions. 
- A custom API was implemented for this data access, with a batch interface, to query the `cardano-db-sync` database, and later custom projections with more a more optimised schema.
- The API schema was implemented from scratch, based on the domain model, due to the lack of an existing standard.

> [!NOTE]
> Since this API was developed, [Cardano DB Sync] has since added a mode to store transactions in CBOR, but at the time of implementation it was not an option.

#### Current Alternatives
[Blockfrost WebSocket Link] is a partial solution for the impedance mismatch problem, however it's not addressing the fragility and standards problem. It's not practical to run this service to satisfy a single-tenant, full node, use case. 

> ## Goals

Reviewer 2: One of our goals (or an open question?) can be supporting clients who won't have a constant live connection in the background. 

I am explicitly talking about mobile devices and push messages/notifications.

Author: https://github.com/will-break-it/wallet-architecture/pull/3/commits/916e2390434a9fed9f51b0def55039ef3e8f4ae1

- Define a standard API that is optimal for fundamental wallet operation. Consider extensibility for optional enhancement.
- Deliver data as a filtered subset of the network state, using native encoding formats, to simplify server implementations and push processing to the edge of the network.
- Foster long term stability of the API when crossing major Cardano protocol upgrade boundaries.
- Facilitate event-driven wallet behavior, minimising the need for polling, and to increase the sophistication of infrastructure management and user-base insights.
- Support script-based wallets
- Minimise the network demand by allowing the server to control the throttling strategy, rather than a pull-based approach like we see in the local chain-sync protocol. 
- API can be applied in a local, single-tenant environment, and a remote multi-tenant deployment, to facilitate a universal approach to wallet supporting an optional full local node for power users. 
- Remove or reduce the need for wallets to perform address discovery

> ## Open Questions

Reviewer 2: There are two more open questions in my mind. I am not sure if they belong here. My comments under the questions are to make the questions more transparent.
- How can we keep re-connections cheap for the client and the server?
     In the case of mobile devices, I suspect there will be plenty of one-second disconnections in a fast-moving car.
- How can we ensure that the first sync of a big wallet is time and network-efficient?
     Considering a busy address, pushing relevant events one by one from the genesis might be time and network-heavy. Is a hybrid solution of sending a snapshot acceptable? Is sending the messages as a batch acceptable?

Author: https://github.com/will-break-it/wallet-architecture/pull/3/commits/80b1097e379cc43b58f39b1086386d97f79205f5
